pkg/controller/elasticsearch/driver/upgrade_pods_selector_test.go

```diff
@@ -0,0 +1,285 @@
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("master-0").isMaster(true).isData(false).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("node-0").isMaster(false).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+				),
+				maxUnavailable: 1,
+				green:          true,
+				podFilter:      nothing,
+			},
+			deleted:                      []string{"node-0"},
+			wantErr:                      false,
+			wantShardsAllocationDisabled: true,
+		},
+		{
+			name: "Do not delete healthy node if not green",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("master-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+				),
+				maxUnavailable: 1,
+				green:          false,
+				podFilter:      nothing,
+			},
+			deleted:                      []string{},
+			wantErr:                      false,
+			wantShardsAllocationDisabled: false,
+		},
+		{
+			name: "Allow deletion of unhealthy node if not green",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(false).needsUpgrade(true).isInCluster(false),
+					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+				),
+				maxUnavailable: 1,
+				green:          false,
+				podFilter:      nothing,
+			},
+			deleted:                      []string{"masters-2"},
+			wantErr:                      false,
+			wantShardsAllocationDisabled: true,
+		},
+		{
+			name: "Do not delete last healthy master",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(false).isInCluster(true),
+					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(false).needsUpgrade(true).isInCluster(false),
+				),
+				maxUnavailable: 1,
+				green:          true,
+				podFilter:      nothing,
+			},
+			deleted:                      []string{"masters-1"},
+			wantErr:                      false,
+			wantShardsAllocationDisabled: true,
+		},
+		{
+			name: "Pod deleted while upgrading",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					newTestPod("masters-0").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-1").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+					newTestPod("masters-2").isMaster(true).isData(true).isHealthy(true).needsUpgrade(true).isInCluster(true),
+				),
+				maxUnavailable: 1,
+				green:          true,
+				podFilter:      byName("masters-2"),
+			},
+			deleted:                      []string{},
+			wantErr:                      true,
+			wantShardsAllocationDisabled: true,
+		},
+		{
+			name: "Do not delete Pods that share some shards",
```

Comment:
It actually tests that `elasticsearch-sample-es-nodes-4` and `elasticsearch-sample-es-nodes-3` which share some shards, are not deleted together.

Reply:
`maxUnavailable: 2` is a bit misleading then, should be `maxUnavailable: 3` (with only 2 effectively upgraded)?

Reply:
If `maxUnavailable` is set to 3 the operator will attempt to delete 3 Pods (which is possible in this test because `elasticsearch-sample-es-nodes-1` does not share some shards with `elasticsearch-sample-es-nodes-4` and `elasticsearch-sample-es-nodes-2`)

I have updated the comments on the test.